feature_engine/encoding/similarity_encoder.py

```diff
@@ -0,0 +1,298 @@
+from typing import List, Optional, Union
+
+import difflib
+import numpy as np
+import pandas as pd
+from sklearn.utils.validation import check_is_fitted
+
+from feature_engine._docstrings.fit_attributes import (
+    _feature_names_in_docstring,
+    _n_features_in_docstring,
+    _variables_attribute_docstring,
+)
+from feature_engine._docstrings.init_parameters import (
+    _ignore_format_docstring,
+    _variables_categorical_docstring,
+)
+from feature_engine._docstrings.methods import _fit_transform_docstring
+from feature_engine._docstrings.substitute import Substitution
+from feature_engine.dataframe_checks import _check_contains_na, check_X
+from feature_engine.encoding.base_encoder import (
+    CategoricalInitMixin,
+    CategoricalMethodsMixin,
+)
+
+
+def _gpm_fast(x1: str, x2: str) -> float:
+    return difflib.SequenceMatcher(None, x1, x2).quick_ratio()
+
+
+_gpm_fast_vec = np.vectorize(_gpm_fast)
+
+
+@Substitution(
+    ignore_format=_ignore_format_docstring,
+    variables=_variables_categorical_docstring,
+    variables_=_variables_attribute_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit_transform=_fit_transform_docstring,
+)
+class StringSimilarityEncoder(CategoricalInitMixin, CategoricalMethodsMixin):
+    """
+    The StringSimilarityEncoder() replaces categorical variables by a set of
+    float variables representing similarity between unique categories in the variable.
+    This new variables will have values in range between 0 and 1, where 0 is the least
+    similar and 1 is the exact match.
+    This encoding is an alternative to OneHotEncoder in the case of poorly
+    defined categorical variables.
+
+    The encoder will create k variables, where k is the number of unique categories.
+
+    The encoder has the additional option to generate similarity variables only for the
+    most popular categories, that is, the categories that are shared by the
+    majority of the observations in the dataset. This behaviour can be specified with
+    the parameter `top_categories`.
+
+    The encoder has the option to specify the behaviour when NaN is present in the
+    variable, see parameter `handle_missing`.
+
+    The encoder will encode only categorical variables by default (type 'object' or
+    'categorical'). You can pass a list of variables to encode. Alternatively, the
+    encoder will find and encode all categorical variables (type 'object' or
+    'categorical').
+
+    With `ignore_format=True` you have the option to encode numerical variables as well.
```

Comment:
This parameter exists for the other encoders because sometimes, numerical variables with numbers like 1,2,3 could be categorical, for example if they represent months, or days, or some sort of internal code.

For this transformer, would it make sense to calculate distances between numbers? When would this be useful?

Reply:
For example in cases where numbers could be represented as a string, for example barcode (substracting them has no sense, but calculating edit distance does). Another one will be ticket column in titanic dataset, sometimes ticket numbers will have different formatting and typos, and this encoding could mitigate that.

But I guess this option could be removed, no hard opinion.

Reply:
Allowing numerical columns also gives us the chance to test the transformer with the check_estimator from sklearn. Maybe we leave it then. 

It would be good to add this example in the user guide.